File: src/engine/graph/persistence/sqlite/writer.ts
import { randomUUID } from "node:crypto";
import type { FileGraphResult } from "../../types.js";
import type { GraphStorage } from "../storage.js";
import type { GraphDatabase } from "./database.js";
import { graphTransaction } from "./transaction.js";

/** Replaces extraction snapshots and invalidates dependent cross-file edges. */
export class SqliteGraphWriter implements Pick<
  GraphStorage,
  "writeFileGraph" | "deleteFileGraph"
> {
  constructor(private readonly database: GraphDatabase) {}

  async deleteFileGraph(
    fileId: string,
    oldEntityIds: readonly string[],
  ): Promise<void> {
    this.replace(
      fileId,
      { nodes: [], edges: [], pendingRefs: [] },
      oldEntityIds,
    );
  }

  /**
   * Accepts file-local extraction output only. Cross-file matches must use
   * applyResolutions(), which retains the original reference. Callers serialize target validation
   * and resolution writeback with file changes.
   * Atomic within SQLite; zvec coordination belongs to the indexing pipeline.
   */
  async writeFileGraph(
    fileId: string,
    result: FileGraphResult,
    oldEntityIds: readonly string[],
  ): Promise<void> {
    this.replace(fileId, result, oldEntityIds);
  }

  private replace(
    fileId: string,
    result: FileGraphResult,
    oldEntityIds: readonly string[],
  ): void {
    if (fileId.trim().length === 0) {
      throw new Error("Graph file ID must not be empty");
    }
    if (
      !Array.isArray(oldEntityIds) ||
      oldEntityIds.some(
        (id) => typeof id !== "string" || id.trim().length === 0,
      )
    ) {
      throw new Error(
        "A complete oldEntityIds array is required; use [] for a new file",
      );
    }
    const targets = [...new Set([fileId, ...oldEntityIds])];
    const localIds = new Set(result.nodes.map((node) => node.id));
    if (
      localIds.size !== result.nodes.length ||
      [...localIds].some((id) => id.trim().length === 0 || id === fileId)
    ) {
      throw new Error(
        "Graph nodes must have unique, non-empty entity IDs distinct from the file ID",
      );
    }
    const owns = (id: string): boolean => id === fileId || localIds.has(id);
    for (const edge of result.edges) {
      if (
        edge.provenance !== "file_local" ||
        !owns(edge.source) ||
        !owns(edge.target)
      ) {
        throw new Error(
          "File snapshots accept only local edges; use applyResolutions for cross-file edges",
        );
      }
    }
    for (const ref of result.pendingRefs) {
      if (!owns(ref.ownerId) || ref.status !== "pending") {
        throw new Error(
          "File snapshots require locally owned pending references",
        );
      }
    }
    const db = this.database.connection;
    graphTransaction(db, () => {
      // Include fileId for file-level imports. Chunk to stay below SQLite's
      // parameter limit for large files; every chunk shares this transaction.
      for (let offset = 0; offset < targets.length; offset += 500) {
        const batch = targets.slice(offset, offset + 500);
        const placeholders = batch.map(() => "?").join(", ");
        db.prepare(
          `UPDATE pending_refs
          SET status = 'pending', token = lower(hex(randomblob(16)))
          WHERE id IN (SELECT ref_id FROM edges WHERE target IN (${placeholders}) AND file_id <> ?)
        `,
        ).run(...batch, fileId);
        db.prepare(`DELETE FROM edges WHERE target IN (${placeholders})`).run(
          ...batch,
        );
      }
      db.prepare("DELETE FROM edges WHERE file_id = ?").run(fileId);
      db.prepare("DELETE FROM pending_refs WHERE file_id = ?").run(fileId);
      const insertEdge = db.prepare(`INSERT INTO edges
        (file_id, kind, source, target, line, column, provenance, metadata)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)`);
      for (const edge of result.edges) {
        insertEdge.run(
          fileId,
          edge.kind,
          edge.source,
          edge.target,
          edge.line,
          edge.column,
          edge.provenance,
          JSON.stringify(edge.metadata),
        );
      }
      const insertRef = db.prepare(`INSERT INTO pending_refs
        (file_id, token, owner_id, ref_name, receiver_name, ref_kind, arity, line, column, status, metadata)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`);
      for (const ref of result.pendingRefs) {
        insertRef.run(
          fileId,
          randomUUID(),
          ref.ownerId,
          ref.refName,
          ref.receiverName,
          ref.refKind,
          ref.arity,
          ref.line,
          ref.column,
          ref.status,
          JSON.stringify(ref.metadata),
        );
      }
    });
  }
}

File: rust/crates/zg-graph-storage/src/writer.rs
use rusqlite::{TransactionBehavior, params, params_from_iter};
use std::collections::HashSet;

use crate::{Error, FileGraph, Provenance, Result, SqliteGraphStorage, nonempty};

impl SqliteGraphStorage {
    /// Atomically replaces one file's local graph and invalidates inbound edges.
    /// `old_entity_ids` must contain **all** pre-update entity IDs from zvec; pass
    /// an empty slice for a new file. Entity IDs in the new snapshot are not stored
    /// in a second node table. Cross-file edges must use `apply_resolutions`.
    ///
    /// # Errors
    /// Rejects invalid ownership, duplicate/empty IDs, non-local edges, invalid
    /// positions, read-only connections, and SQLite failures. Failures roll back.
    pub fn write_file_graph(
        &mut self,
        file_id: &str,
        graph: &FileGraph,
        old_entity_ids: &[String],
    ) -> Result<()> {
        validate(file_id, graph, old_entity_ids)?;
        let tx = self
            .connection
            .transaction_with_behavior(TransactionBehavior::Immediate)?;
        let mut targets: Vec<&str> = old_entity_ids.iter().map(String::as_str).collect();
        targets.push(file_id);
        targets.sort_unstable();
        targets.dedup();
        for chunk in targets.chunks(500) {
            let placeholders = vec!["?"; chunk.len()].join(",");
            // Retain the original reference for the next resolution pass.
            tx.execute(
                &format!("UPDATE pending_refs SET status = 'pending'
                  WHERE id IN (SELECT ref_id FROM edges WHERE target IN ({placeholders}) AND file_id <> ?)"),
                params_from_iter(chunk.iter().copied().chain(std::iter::once(file_id))),
            )?;
            tx.execute(
                &format!("DELETE FROM edges WHERE target IN ({placeholders})"),
                params_from_iter(chunk.iter().copied()),
            )?;
        }
        tx.execute("DELETE FROM edges WHERE file_id = ?", [file_id])?;
        tx.execute("DELETE FROM pending_refs WHERE file_id = ?", [file_id])?;
        {
            let mut insert = tx.prepare(
                "INSERT INTO edges
                (file_id, kind, source, target, line, column, provenance, metadata)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
            )?;
            for edge in &graph.edges {
                insert.execute(params![
                    file_id,
                    edge.kind.as_str(),
                    edge.source,
                    edge.target,
                    edge.line,
                    edge.column,
                    edge.provenance.as_str(),
                    serde_json::to_string(&edge.metadata)?
                ])?;
            }
            let mut insert = tx.prepare("INSERT INTO pending_refs
                (file_id, owner_id, ref_name, receiver_name, ref_kind, arity, line, column, status, metadata)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'pending', ?)")?;
            for reference in &graph.pending_refs {
                insert.execute(params![
                    file_id,
                    reference.owner_id,
                    reference.ref_name,
                    reference.receiver_name,
                    crate::EdgeKind::from(reference.ref_kind).as_str(),
                    reference.arity,
                    reference.line,
                    reference.column,
                    serde_json::to_string(&reference.metadata)?
                ])?;
            }
        }
        tx.commit()?;
        Ok(())
    }

    /// Deletes owned rows and invalidates incoming references, including imports
    /// targeting the file itself. Repeating deletion is safe.
    ///
    /// # Errors
    /// Rejects empty IDs and SQLite failures. The complete mutation is atomic.
    pub fn delete_file_graph(&mut self, file_id: &str, old_entity_ids: &[String]) -> Result<()> {
        self.write_file_graph(file_id, &FileGraph::default(), old_entity_ids)
    }
}

fn validate(file_id: &str, graph: &FileGraph, old_entity_ids: &[String]) -> Result<()> {
    nonempty(file_id)?;
    for id in old_entity_ids {
        nonempty(id)?;
    }
    let mut local = HashSet::new();
    for id in &graph.entity_ids {
        nonempty(id)?;
        if id == file_id || !local.insert(id.as_str()) {
            return Err(Error::InvalidInput(
                "entity IDs must be unique and distinct from the file ID",
            ));
        }
    }
    let owns = |id: &str| id == file_id || local.contains(id);
    for edge in &graph.edges {
        if edge.provenance != Provenance::FileLocal || !owns(&edge.source) || !owns(&edge.target) {
            return Err(Error::InvalidInput(
                "file snapshots require local edges; use apply_resolutions for cross-file edges",
            ));
        }
        if edge.line == Some(0) {
            return Err(Error::InvalidInput("edge lines must be one-based"));
        }
    }
    for reference in &graph.pending_refs {
        nonempty(&reference.ref_name)?;
        if !owns(&reference.owner_id) || reference.line == 0 {
            return Err(Error::InvalidInput(
                "pending refs require local ownership and one-based lines",
            ));
        }
    }
    Ok(())
}
